clippy_lints/src/matches.rs

```diff
@@ -712,6 +821,69 @@ fn check_wild_in_or_pats(cx: &LateContext<'_, '_>, arms: &[Arm<'_>]) {
     }
 }
 
+fn check_match_single_binding(cx: &LateContext<'_, '_>, ex: &Expr<'_>, arms: &[Arm<'_>], expr: &Expr<'_>) {
+    if in_macro(expr.span) || arms.len() != 1 || is_refutable(cx, arms[0].pat) {
+        return;
+    }
+    let matched_vars = ex.span;
+    let bind_names = arms[0].pat.span;
+    let match_body = remove_blocks(&arms[0].body);
+    let mut snippet_body = if match_body.span.from_expansion() {
+        Sugg::hir_with_macro_callsite(cx, match_body, "..").to_string()
+    } else {
+        snippet_block(cx, match_body.span, "..").to_owned().to_string()
+    };
+
+    // Do we need to add ';' to suggestion ?
+    if_chain! {
+        if let ExprKind::Block(block, _) = &arms[0].body.kind;
+        if block.stmts.len() == 1;
+        if let StmtKind::Semi(s) = block.stmts.get(0).unwrap().kind;
```

Comment:
You don't ever need a semicolon in this case, since this is either
```rust
{
    x;
    y
}
```
or
```rust
{
    x;
}
```
In both cases you don't need a semicolon. 

Instead of this check, you can exchange the `s.kind` below with `match_body.kind`. This also makes more sense, since the suggestion is build with the `match_body`, not with the `arms[0].body`, so whether you have to add a semicolon depends only on the `match_body`.

Reply:
Ok I just understand what's going on.
I was confused because the `println!` macro was tagged as block, so I needed to manage this case just after matching the `match_body.kind`.